src/shutdown.rs

```diff
@@ -0,0 +1,253 @@
+
+    /// When a Source allows this to go out of scope it informs the global shutdown coordinator that
+    /// this Source's local shutdown process is complete.
+    /// Optional only so that `poll()` can move the handle out and return it.
+    shutdown_complete: Option<ShutdownCompleteHandle>,
+}
+
+impl Future for ShutdownSignal {
+    type Item = ShutdownCompleteHandle;
+    type Error = ();
+    fn poll(&mut self) -> Result<Async<Self::Item>, Self::Error> {
+        match self.begin_shutdown.poll() {
+            Ok(Async::Ready(_)) => Ok(Async::Ready(self.shutdown_complete.take().unwrap())),
+            Ok(Async::NotReady) => Ok(Async::NotReady),
+            Err(_) => Err(()),
+        }
+    }
+}
+
+impl ShutdownSignal {
+    pub fn new(begin_shutdown: Tripwire, shutdown_complete: Trigger) -> Self {
+        Self {
+            begin_shutdown,
+            shutdown_complete: Some(ShutdownCompleteHandle::new(shutdown_complete)),
+        }
+    }
+
+    /// Only for testing.
+    pub fn noop() -> Self {
+        let (trigger, tripwire) = Tripwire::new();
+        Self {
+            begin_shutdown: tripwire,
+            shutdown_complete: Some(ShutdownCompleteHandle::new(trigger)),
+        }
+    }
+}
+
+pub struct ShutdownCoordinator {
+    shutdown_begun_triggers: HashMap<String, Trigger>,
+    shutdown_force_triggers: HashMap<String, Trigger>,
+    shutdown_complete_tripwires: HashMap<String, Tripwire>,
+}
+
+impl ShutdownCoordinator {
+    pub fn new() -> Self {
+        Self {
+            shutdown_begun_triggers: HashMap::new(),
+            shutdown_complete_tripwires: HashMap::new(),
+            shutdown_force_triggers: HashMap::new(),
+        }
+    }
+
+    /// Creates the necessary Triggers and Tripwires for coordinating shutdown of this Source and
+    /// stores them as needed.  Returns the ShutdownSignal for this Source as well as a Tripwire
+    /// that will be notified if the Source should be forcibly shut down.
+    pub fn register_source(&mut self, name: &str) -> (ShutdownSignal, Tripwire) {
+        let (shutdown_begun_trigger, shutdown_begun_tripwire) = Tripwire::new();
+        let (force_shutdown_trigger, force_shutdown_tripwire) = Tripwire::new();
+        let (shutdown_complete_trigger, shutdown_complete_tripwire) = Tripwire::new();
+
+        self.shutdown_begun_triggers
+            .insert(name.to_string(), shutdown_begun_trigger);
+        self.shutdown_force_triggers
+            .insert(name.to_string(), force_shutdown_trigger);
+        self.shutdown_complete_tripwires
+            .insert(name.to_string(), shutdown_complete_tripwire);
+
+        let shutdown_signal =
+            ShutdownSignal::new(shutdown_begun_tripwire, shutdown_complete_trigger);
+
+        (shutdown_signal, force_shutdown_tripwire)
+    }
+
+    /// Takes ownership of all internal state for the given source from another ShutdownCoordinator.
+    pub fn takeover_source(&mut self, name: &str, other: &mut Self) {
```

Comment:
Do we need to worry about a name collision here (i.e. if we already have a source with this name that the `insert`s will clobber)?

Reply:
Done